bin/Main.re

```diff
@@ -1,27 +1,49 @@
 open Odiff.ImageIO;
 open Odiff.Diff;
 
-let getIOModule = filename =>
+let getTypeFromFilename = filename =>
   Filename.extension(filename)
   |> (
     fun
-    | ".png" => ((module ODiffIO.Png.IO): (module ImageIO))
+    | ".png" => `png
     | ".jpg"
-    | ".jpeg" => ((module ODiffIO.Jpg.IO): (module ImageIO))
-    | ".bmp" => ((module ODiffIO.Bmp.IO): (module ImageIO))
-    | ".tiff" => ((module ODiffIO.Tiff.IO): (module ImageIO))
+    | ".jpeg" => `jpg
+    | ".bmp" => `bmp
+    | ".tiff" => `tiff
     | f => failwith("This format is not supported: " ++ f)
   );
 
+let getIOModule =
+  fun
+  | `png => ((module ODiffIO.Png.IO): (module ImageIO))
+  | `jpg => ((module ODiffIO.Jpg.IO): (module ImageIO))
+  | `bmp => ((module ODiffIO.Bmp.IO): (module ImageIO))
+  | `tiff => ((module ODiffIO.Tiff.IO): (module ImageIO));
+
 type diffResult('output) = {
   exitCode: int,
   diff: option('output),
 };
 
+let readFromStdin = () => {
+  /* We use 65536 because that is the size of OCaml's IO buffers. */
+  let chunk_size = 65536;
+  let buffer = Buffer.create(chunk_size);
+  let rec loop = () => {
+    Buffer.add_channel(buffer, stdin, chunk_size);
+    loop();
+  };
+  try(loop()) {
+  | End_of_file => Buffer.contents(buffer)
+  };
```

Comment:
We are reading from stdin until there is no more data provided.
If we wanted to support bot base and compare image as buffer inputs, we would need to implement a check for a separator in here.

Reply:
Yeah, that is basically why I didn't implement this from scratch. Looks incredibly tough to correctly read stdin and avoid memory copying.

Reply:
I mean the performance is not bad. The only issue is knowing when the first image ends and the second one starts.
I will need to see what I can come up with to solve this problem.